### PR TITLE
Prevent Gdrive update permissions to change domain

### PR DESCRIPTION
**Pull request by PopDaph**
Following https://github.com/dust-tt/dust/pull/989, this one is for Gdrive. 
We want to prevent users from changing the Gdrive when they are trying to update its permissions (adding or removing folders).

I couldn't find an easy way to retrieve the Google Project id from the Google Drive client API. I might have missed it, but I did not wanted to spend too much time on this, so I decided to opt for a workaround that I think should work. Of course, dear reviewers, if I missed it, I'd be happy to change that logic 🙇🏻‍♀️ 

I'm retrieving the user who initiated the permission to the grive and comparing its domain from her email address with the domain from the email address who initiated the new permission. It should cover all cases as we only handle shared gdrive

WDYT? 😬

### CODE REVIEW

**Comment by spolu** [comment]
Surprised that we couldn’t get an ID from the “about”.

this seems a good enough remediation even if I presume exceptions will exist with multi domain organisations etc…

**Comment by PopDaph** [comment]
> Surprised that we couldn’t get an ID from the “about”.

Yes, me too! 
Here's a script I used for testing if any reviewer wants to check by themselves easily: 


```
import { google } from "googleapis";

import {
  getAuthObject,
  getDriveClient,
} from "@connectors/connectors/google_drive/temporal/activities";
import { Connector } from "@connectors/lib/models";

async function main() {
  // Getting the connection id
  const connectorId = 12;
  const c = await Connector.findOne({
    where: {
      id: connectorId,
    },
  });
  if (!c) {
    console.log("Connector not found");
    return;
  }

  // ABOUT ROUTE
  const driveClient = await getDriveClient(c.connectionId);
  driveClient.drives.list({}, (err, res) => {
    if (err) return console.error("There was an error:", err);
    const drives = res.data.drives;
    if (drives) {
      console.log("Drives:");
      drives.forEach((drive) => console.log(`${drive.name} (${drive.id})`));
    } else {
      console.log("No drives found.");
    }
  });

  const userFromDrive = await driveClient.about.get({ fields: "*" });
  console.log(userFromDrive.data); // there's a lot here but nothing useful
  console.log(userFromDrive.data.user?.emailAddress);

  // ADMIN ? WE DON'T HAVE ACCESS TO THIS API
  // const admin = google.admin({
  //   version: "directory_v1",
  //   auth: await getAuthObject(c.connectionId),
  // });

  // admin.users.get(
  //   {
  //     userKey: "your-email@your-domain.com", // replace with the user's email
  //   },
  //   (err, res) => {
  //     if (err) return console.error("There was an error:", err);

  //     console.log("Organization ID:", res.data.customerId);
  //   }
  // );

  // THIS WORKS FOR OUR ACCOUNT BUT NOT FOR A USER ACCOUNT
  const test = await google.auth.getProjectId();
  console.log(test);

  const auth = await getAuthObject(c.connectionId);
  console.log(auth.projectId); // that's undefined
}

main().catch((err) => {
  console.log(err);
  process.exit(1);
});
```

File: connectors/src/connectors/google_drive/index.ts [hunk 1]
```diff
@@ -92,6 +93,85 @@ export async function createGoogleDriveConnector(
...
+    },
+  });
+  if (!c) {
+    logger.error({ connectorId }, "Connector not found");
+    return new Err({
+      error: {
+        message: "Connector not found",
+        type: "connector_not_found",
+      },
+    } as ConnectorsAPIErrorResponse);
```
Comment (fontanierh) [inline connectors/src/connectors/google_drive/index.ts:120]: Is the `as` needed here ? If so, why ?

File: connectors/src/connectors/google_drive/index.ts [hunk 2]
```diff
@@ -92,6 +93,85 @@ export async function createGoogleDriveConnector(
...
+    const currentDriveUserDomain = currentUserEmail.split("@")[1];
+
+    const newDriveClient = await getDriveClient(connectionId);
+    const newDriveUser = await newDriveClient.about.get({
+      fields: "user",
+    });
+    const newDriveUserEmail = newDriveUser.data?.user?.emailAddress || "";
+    const newDriveUserDomain = newDriveUserEmail.split("@")[1];
+
+    if (!currentDriveUserDomain || !newDriveUserDomain) {
```
Comment (fontanierh) [inline connectors/src/connectors/google_drive/index.ts:142]: Do we get the email for all users ?

**Review by fontanierh — DISMISSED** [review]
Small comments  / question but overall LVGTM I think this will do the trick in 90%+ of situations

File: connectors/src/connectors/google_drive/index.ts (hunk 1 above)
Comment (PopDaph) [reply connectors/src/connectors/google_drive/index.ts:120]: Thank you for noticing, It's not, I removed it from all other connectors and forgot for this one 🤦🏻‍♀️

File: connectors/src/connectors/google_drive/index.ts (hunk 2 above)
Comment (PopDaph) [reply connectors/src/connectors/google_drive/index.ts:142]: For the two accounts I tested yes 😬 
But indeed I just (finally) found the valid doc and 

About returns this: https://developers.google.com/drive/api/reference/rest/v3/User?hl=fr
And the User has an optional address: https://developers.google.com/drive/api/reference/rest/v3/User?hl=fr

> Uniquement en sortie. Adresse e-mail de l'utilisateur. Elle peut ne pas être présente dans certains cas si l'utilisateur n'a pas rendu son adresse e-mail visible par le demandeur.

If that happens then this specific user will be blocked and won't be able to update the permission. I guess that's okay enough?

Comment (lasryaric) [reply connectors/src/connectors/google_drive/index.ts:142]: I see one pontential issue:
if someone signs up via lets say github, I think we never request the scope to get their email in the Google Drive Oauth process. Testing is needed to see what happens in this case.

**Review by lasryaric — COMMENTED** [review]
At least needs testing when logging in via Github.

File: connectors/src/connectors/google_drive/index.ts (hunk 2 above)
Comment (PopDaph) [reply connectors/src/connectors/google_drive/index.ts:142]: Even if you signed up by Github, if you want to setup a Grive connector you'll have to oAuth with your Google account right? I don't see why we wouldn't get the email this way, can you explain?

Comment (PopDaph) [reply connectors/src/connectors/google_drive/index.ts:142]: I just tested this and it works: 
- Create a new workspace attached to my Github account (@gmail.com)
- Create a gdrive datasource for casquedelumiere
- Edit permissions: Try to add another folder for the same gdrive -> works
- Edit permissions: Select dust workspace -> blocked as expected.

Comment (PopDaph) [reply connectors/src/connectors/google_drive/index.ts:142]: What do you mean by "Disconnected from Dust"? 🙏🏻

**Pull request merged**